File: modules/innertube/android/src/main/java/expo/modules/innertube/InnerTubeModule.kt
package expo.modules.innertube

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class InnerTubeModule : Module() {
  override fun definition() = ModuleDefinition {
    // The module will be accessible from `requireNativeModule('InnerTube')` in JavaScript.
    Name("InnerTube")

    // Runs on JavaScript thread.
    Function("hello") { "Hello from InnerTube - synchronously!?" }

    // Runs on background thread.
    AsyncFunction("someAsyncFunction") { value: String -> "Received value from JavaScript: $value" }
  }
}

File: modules/innertube/src/InnerTubeModule.ts
import { NativeModule, requireNativeModule } from "expo"

import type { InnerTubeModuleEvents } from "./InnerTube.types"

declare class InnerTubeModule extends NativeModule<InnerTubeModuleEvents> {
  hello(): string
  someAsyncFunction(value: string): Promise<string>
}

// This call loads the native module object from the JSI.
export default requireNativeModule<InnerTubeModule>("InnerTube")
